### PR TITLE
[NEW] New avatar storage types

### PR DESCRIPTION
**Pull request by sampaiodiego**
@RocketChat/core 

This PR will change avatar storage type to use the same as file uploads, so all different storage types will available for both:

- [x] File System
- [x] GridFS
- [x] Amazon S3
- [x] Google Cloud Storage
- [x] Server upload for File System
- [x] Server upload for GridFS
- [x] Server upload for Amazon S3
- [x] Server upload for Google Cloud Storage
- [x] Remove references of `RocketChatFileAvatarInstance`
- [x] Backward compatibility for mobile apps
- [x] Option to upload to S3 passing through server
- [x] Option to upload to GCS passing through server
- [x] Migration
- [x] Migrate old Avatars to new format
- [ ] Option to migrate files between stores

### Not important
- [ ] Create a setting to prevent showing avatars to not logged in users
- [ ] CDN option for S3 Server

Closes https://github.com/RocketChat/Rocket.Chat/issues/3480

## Doc
There will be 4 types of stores:
* AmazonS3
* GoogleCloudStorage
* GridFS
* FileSystem

And 2 types of uploads
* Avatars
* Uploads

The upload store is a combination of both (store and upload types), like `AmazonS3:Uploads`, but to prevent the need to know the current active store there are 2 fake stores that point to current ones:
* Avatars
* Uploads

If you upload pointing to `Uploads` the file will be saved to the current store as a upload type.

### Backwards compatibility
The old store names `fileSystem` and `gridFS` are still supported but deprecated.
The direct uploads to `s3` and `googleCloudStorage` are still supported but deprecated.

### CODE REVIEW

File: packages/rocketchat-file-upload/server/lib/FileUpload.js [hunk 1]
```diff
@@ -1,43 +1,264 @@
...
+	},
+
+	defaultUploads() {
+		return {
+			collection: RocketChat.models.Uploads.model,
+			filter: new UploadFS.Filter({
+				onCheck: FileUpload.validateFileUpload
+			}),
+			getPath(file) {
+				return `${ RocketChat.settings.get('uniqueID') }/uploads/${ file.rid }/${ file.userId }/${ file._id }`;
```
Comment (sampaiodiego) [inline packages/rocketchat-file-upload/server/lib/FileUpload.js]: `file._id` is `undefined`.. that means all files are being saved as `undefined` (and being all replaced)..

File: packages/rocketchat-file-upload/lib/FileUploadBase.js [hunk 2]
```diff
@@ -15,10 +15,13 @@ UploadFS.config.defaultStorePermissions = new UploadFS.StorePermissions({
 
 
 FileUploadBase = class FileUploadBase {
-	constructor(meta, file) {
+	constructor(store, meta, file) {
 		this.id = Random.id();
 		this.meta = meta;
 		this.file = file;
+
+		console.log(store.options.name, {meta, file});
```
Comment (sampaiodiego) [inline packages/rocketchat-file-upload/lib/FileUploadBase.js]: left `console.log`

File: packages/rocketchat-file-upload/client/lib/fileUploadHandler.js [hunk 3]
```diff
@@ -1,10 +1,34 @@
...
+
+	if (store) {
+		return new FileUploadBase(store, meta, file);
+	} else {
+		console.error('Invalid file store', directive);
 	}
 };
+
+Tracker.autorun(function() {
+	document.cookie = `rc_uid=${ escape(Meteor.userId()) }; path=/`;
```
Comment (sampaiodiego) [inline packages/rocketchat-file-upload/client/lib/fileUploadHandler.js]: should we set the cookies only when logged in?

File: packages/rocketchat-file-upload/server/lib/FileUpload.js (hunk 1 above)
Comment (rodrigok) [inline packages/rocketchat-file-upload/server/lib/FileUpload.js]: Fixed

**Comment by rodrigok** [comment]
@sampaiodiego All fixed

**Pull request merged**